### PR TITLE
[FLINK-16687][python] Support Python UDF in Java Correlate

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will support Python UDF in Java Correlate*

## Brief change log

  - *Add the logic of extracting Python UDF in PythonCorrelateSplitRule *

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit test of testJavaTableFunctionWithPythonCalc in PythonCorrelateSplitRuleTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCorrelateSplitRule.java [hunk 1]
```diff
@@ -101,13 +102,13 @@ public boolean matches(RelOptRuleCall call) {
 		RelDataType rowType,
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCorrelateSplitRule.java:102]: Typo, should be `createNewFieldNames`

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCorrelateSplitRule.java [hunk 2]
```diff
@@ -60,7 +60,7 @@ private FlinkLogicalTableFunctionScan createNewScan(
 		FlinkLogicalTableFunctionScan scan,
 		ScalarFunctionSplitter splitter) {
 		RexCall rightRexCall = (RexCall) scan.getCall();
-		// extract java funcs.
+		// extract Java funcs of Python TableFunction or Python funcs of Java TableFunction.
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCorrelateSplitRule.java:63]: `extract Java funcs from Python TableFunction or Python funcs from Java TableFunction.`

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCorrelateSplitRule.java [hunk 3]
```diff
@@ -185,36 +186,59 @@ private FlinkLogicalCalc createTopCalc(
 				rexBuilder));
 	}
 
+	private ScalarFunctionSplitter createScalarFunctionSplitter(
+		int primitiveLeftFieldCount,
+		ArrayBuffer<RexCall> extractedRexCalls,
+		boolean isJavaTableFunction) {
+		if (isJavaTableFunction) {
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCorrelateSplitRule.java:193]: Can be simplified as:
```
		return new ScalarFunctionSplitter(
			primitiveLeftFieldCount,
			extractedRexCalls,
			isJavaTableFunction ? x -> PythonUtil.isPythonCall(x, null) : PythonUtil::isNonPythonCall
		);
```

**Comment by HuangXingBo** [comment]
Thanks a lot for @hequn8128 review, I have addressed the comments at the latest commit.

**Pull request merged**